exercises/circular-buffer/canonical-data.json

```diff
@@ -0,0 +1,296 @@
+{
+  "#": [
+    "In general, these circular buffers are expected to be stateful,",
+    "and each language will operate on them differently.",
+    "Tests tend to perform a series of operations, some of which expect a certain result.",
+    "As such, this common test suite can only say in abstract terms what should be done.",
+    "",
+    "Tests will contain a number of operations. The operation will be specified in the `operation` key.",
+    "Based on the operation, other keys may be present.",
+    "read: Reading from the buffer should produce the item located at `expected`.",
+    "write: Writing the item located at `item` should succeed.",
+    "readShouldFail: Reading from the buffer should fail.",
+    "writeShouldFail: Writing the item located at `item` should fail.",
+    "overwrite: Write the item located at `item` into the buffer, removing the oldest item if necessary.",
+    "clear: Clear the buffer.",
+    "",
+    "Finally, note that all values are integers.",
+    "If your language contains generics, you may consider allowing buffers to contain other types.",
+    "Tests for that are not included here.",
+    ""
+  ],
+  "cases": [
+    {
+      "description": "reading empty buffer should fail",
+      "capacity": 1,
+      "operations": [
+        {
+          "operation": "readShouldFail"
```

Comment:
Okay, I think I understand.

A few things give me pause about the `"expected": false` to signal failure on `read`:

* Not a concern: ~~now we can't have a buffer of booleans and write false into it and expect to read a false out!~~ (not relevant for this JSON file since it only deals with integers)
* The explanation of read gets a little gnarly. It becomes "read: Reading from the buffer should produce the item located at `expected`, unless `expected` is false, in which case the read should fail."
* Statically-typed languages reading this JSON file may have more trouble dealing with the fact that the value at `expected` might be either a boolean or an integer.

Okay, what if it's something like this: for a successful read of 1:

```json
{
  operation: "read",
  item: "1",
  expected: true,
}
```

and for a failing read:

```json
{
  operation: "read",
  expected: false,
}
```

this way, `expected` is a boolean for all cases.

I'm not convinced I'm happy with the name `expected` in this usage, since now it doesn't make clear that the main thing we want to check is the value read, rather than whether it was successful. What if `expected` got renamed to `success` or `shouldSucceed`?

I'm not sure if this is also part of the suggestion, but I would feel odd about adding an `expected`: key to `overwrite` and `clear`, since having it there implies that it can be false, and I don't think we'll ever have a test that expects either of these two to fail (well, maybe overwrite on a size-0 buffer, but I think I'd rather argue that size-0 buffers should just not get created)

I am having a hard time with the `"expected": false` for reads, and I'm not sure that moving the item to a separate key makes things better. Any way to make this work?